File: services/django-ddd/src/domain/repositories/user_repository.py
"""UserRepository 契約。

ここには ABC のみを置く。実装は infrastructure/django_orm/repositories/ にある。
戻り値は必ず「エンティティ」であること。Django Model や QuerySet を返さない。
それを返した瞬間に、ORM の都合が domain を越えて漏れ出す。
"""
from __future__ import annotations

from abc import ABC, abstractmethod

from domain.entities.user import User
from domain.value_objects.email import Email
from domain.value_objects.user_id import UserId


class UserRepository(ABC):
    @abstractmethod
    def find_by_id(self, user_id: UserId) -> User | None:
        """ID でユーザーを取得する。存在しなければ None。"""

    @abstractmethod
    def find_by_email(self, email: Email) -> User | None:
        """Email でユーザーを取得する。存在しなければ None。"""

    @abstractmethod
    def save(self, user: User) -> User:
        """ユーザーを永続化する（新規・更新の両方）。"""

File: services/laravel-clean/app/Repositories/UserRepository.php
<?php

declare(strict_types=1);

namespace App\Repositories;

use App\DataTransferObjects\UserDto;
use App\Models\User;

/**
 * UserRepositoryInterface の Eloquent 実装。
 *
 * Model を use してよい唯一の層。
 * **返す直前に必ず Model → Dto へ変換する。**
 */
final readonly class UserRepository implements UserRepositoryInterface
{
    public function findById(string $id): ?UserDto
    {
        $row = User::query()->find($id);

        return $row === null ? null : $this->toDto($row);
    }

    public function findByEmail(string $email): ?UserDto
    {
        $row = User::query()->where('email', $email)->first();

        return $row === null ? null : $this->toDto($row);
    }

    public function save(UserDto $user): UserDto
    {
        $row = User::query()->updateOrCreate(
            ['id' => $user->id],
            [
                'email' => $user->email,
                'display_name' => $user->displayName,
                'is_active' => $user->isActive,
            ],
        );

        return $this->toDto($row);
    }

    /** Model → Dto 変換。この境界で ORM の都合を断ち切る。 */
    private function toDto(User $row): UserDto
    {
        return new UserDto(
            id: $row->id,
            email: $row->email,
            displayName: $row->display_name,
            isActive: $row->is_active,
        );
    }
}
